File: examples/dekita/.claude/hooks/similar-pattern-search.py
#!/usr/bin/env python3
"""PRマージ後にコードベース内の類似パターンを検索し修正漏れを防ぐ。

Why:
    共通パターン（json.dumps等）を修正する際、同様のパターンが他ファイルに
    存在すると修正漏れが発生する。マージ後に自動検索して警告する。

What:
    - PRマージ成功後（PostToolUse:Bash）に発火
    - PR diffから関数呼び出しパターンを抽出
    - 変更されたファイル以外で同パターンを検索
    - 見つかった場合はsystemMessageで通知

Remarks:
    - 非ブロック型（情報提供のみ）
    - duplicate-issue-checkはIssue重複、本フックはコードパターン重複
    - 一般的すぎる関数（print, len等）はCOMMON_FUNCTIONSで除外

Changelog:
    - silenvx/dekita#2103: フック追加（Issue #2054/2065の再発防止）
"""

from __future__ import annotations

import json
import re
import subprocess
import sys
from pathlib import Path

HOOKS_DIR = Path(__file__).parent
sys.path.insert(0, str(HOOKS_DIR))
from lib.constants import TIMEOUT_LIGHT, TIMEOUT_MEDIUM
from lib.execution import log_hook_execution
from lib.hook_input import get_exit_code, get_tool_result
from lib.repo import is_merge_success
from lib.session import parse_hook_input

# 検索結果の最大表示件数
MAX_RESULTS = 5

# 検索対象から除外するパターン
EXCLUDE_PATTERNS = [
    "*.pyc",
    "__pycache__",
    "node_modules",
    ".git",
    "*.min.js",
    "*.min.css",
    "pnpm-lock.yaml",
    "package-lock.json",
]

# 抽出する関数呼び出しパターン（よくある修正対象）
# 否定後読みでdef/class定義を除外し、関数呼び出しのみをマッチ
# Note: \bは否定後読み内では使用しない（ゼロ幅アサーションのため）
FUNCTION_PATTERN = re.compile(
    r"(?<!def )(?<!class )\b([a-zA-Z_][a-zA-Z0-9_]*(?:\.[a-zA-Z_][a-zA-Z0-9_]*)*)\s*\(",
)

# 除外する一般的すぎる関数名
COMMON_FUNCTIONS = {
    "print",
    "len",
    "str",
    "int",
    "float",
    "list",
    "dict",
    "set",
    "tuple",
    "range",
    "enumerate",
    "zip",
    "map",
    "filter",
    "sorted",
    "reversed",
    "open",
    "type",
    "isinstance",
    "hasattr",
    "getattr",
    "setattr",
    "self",
    "super",
    "return",
    "if",
    "for",
    "while",
    "with",
    "assert",
    "raise",
    "except",
    "import",
    "from",
    "class",
    "def",
    "async",
    "await",
    "lambda",
    # 非常に一般的な関数
    "get",
    "add",
    "remove",
    "pop",
    "append",
    "extend",
    "update",
    "items",
    "keys",
    "values",
    "join",
    "split",
    "strip",
    "replace",
    "format",
    "lower",
    "upper",
    "startswith",
    "endswith",
    "find",
    "index",
    "count",
}


def is_pr_merge_command(command: str) -> bool:
    """Check if the command is a PR merge command."""
    return "gh pr merge" in command


def extract_pr_number(command: str) -> int | None:
    """Extract PR number from merge command or current branch PR."""
    # Match patterns like: gh pr merge 123, gh pr merge #123, gh pr merge --squash 789
    match = re.search(r"gh\s+pr\s+merge\s+.*?#?(\d+)", command)
    if match:
        return int(match.group(1))

    # If no PR number in command, get PR for current branch
    try:
        result = subprocess.run(
            ["gh", "pr", "view", "--json", "number"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_LIGHT,
        )
        if result.returncode == 0:
            data = json.loads(result.stdout)
            return data.get("number")
    except (subprocess.TimeoutExpired, json.JSONDecodeError, FileNotFoundError):
        # gh CLI not available or network error - return None
        pass

    return None


def get_pr_diff(pr_number: int) -> str | None:
    """Get the diff of a PR."""
    try:
        result = subprocess.run(
            ["gh", "pr", "diff", str(pr_number)],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
        )
        if result.returncode == 0:
            return result.stdout
    except (subprocess.TimeoutExpired, FileNotFoundError):
        # gh CLI not available or timeout - return None
        pass
    return None


def get_changed_files(pr_number: int) -> list[str]:
    """Get list of files changed in the PR."""
    try:
        result = subprocess.run(
            ["gh", "pr", "view", str(pr_number), "--json", "files"],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_LIGHT,
        )
        if result.returncode == 0:
            data = json.loads(result.stdout)
            files = data.get("files", [])
            return [f.get("path", "") for f in files if f.get("path")]
    except (subprocess.TimeoutExpired, json.JSONDecodeError, FileNotFoundError):
        # gh CLI not available or error - return empty list
        pass
    return []


def extract_function_patterns(diff: str) -> set[str]:
    """Extract function call patterns from diff.

    Focuses on added/modified lines (lines starting with +).
    """
    patterns = set()

    for line in diff.split("\n"):
        # Focus on added/modified lines
        if not line.startswith("+"):
            continue
        # Skip diff headers
        if line.startswith("+++"):
            continue

        # Extract function calls
        matches = FUNCTION_PATTERN.findall(line)
        for match in matches:
            # Skip common functions
            func_name = match.split(".")[-1]  # Get last part for method calls
            if func_name.lower() not in COMMON_FUNCTIONS:
                patterns.add(match)

    return patterns


def search_pattern_in_codebase(pattern: str, exclude_files: list[str]) -> list[dict]:
    """Search for a pattern in the codebase using ripgrep."""
    results = []

    # Build exclude arguments
    exclude_args = []
    for excl in EXCLUDE_PATTERNS:
        exclude_args.extend(["-g", f"!{excl}"])

    # Exclude changed files
    for f in exclude_files:
        exclude_args.extend(["-g", f"!{f}"])

    try:
        # Escape pattern for regex
        escaped_pattern = re.escape(pattern) + r"\s*\("

        result = subprocess.run(
            [
                "rg",
                "--line-number",
                "--no-heading",
                "--max-count",
                "10",  # Limit matches per file
                *exclude_args,
                escaped_pattern,
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            cwd=Path.cwd(),
        )

        if result.returncode == 0 and result.stdout.strip():
            for line in result.stdout.strip().split("\n")[:MAX_RESULTS]:
                # Parse rg output: file:line:content
                parts = line.split(":", 2)
                if len(parts) >= 3:
                    results.append(
                        {
                            "file": parts[0],
                            "line": parts[1],
                            "content": parts[2].strip()[:80],  # Truncate
                        }
                    )
    except (subprocess.TimeoutExpired, FileNotFoundError):
        # rg not available or timeout - return empty
        pass

    return results


def format_info_message(pattern_results: dict[str, list[dict]]) -> str:
    """Format the informational message."""
    lines = [
        "🔍 **修正漏れの可能性があります**",
        "",
        "PRで変更されたパターンと類似のコードが他のファイルにあります:",
        "",
    ]

    for pattern, results in pattern_results.items():
        lines.append(f"**`{pattern}`**:")
        for r in results[:3]:  # Show max 3 per pattern
            lines.append(f"  - `{r['file']}:{r['line']}` - {r['content']}")
        if len(results) > 3:
            lines.append(f"  - ... 他 {len(results) - 3} 件")
        lines.append("")

    lines.append("同様の修正が必要ないか確認してください。")

    return "\n".join(lines)


def main() -> None:
    """Main hook logic."""
    result = {"decision": "approve"}

    try:
        input_data = parse_hook_input()
        if not input_data:
            print(json.dumps(result))
            return

        tool_name = input_data.get("tool_name", "")
        if tool_name != "Bash":
            print(json.dumps(result))
            return

        tool_input = input_data.get("tool_input", {})
        command = tool_input.get("command", "")

        if not is_pr_merge_command(command):
            print(json.dumps(result))
            return

        tool_output = input_data.get("tool_output", "")
        # Issue #2203: Use get_exit_code() for consistent default value
        # Use get_tool_result() to handle both tool_result and tool_response
        tool_result = get_tool_result(input_data) or {}
        exit_code = get_exit_code(tool_result)

        if not is_merge_success(exit_code, tool_output, command):
            print(json.dumps(result))
            return

        pr_number = extract_pr_number(command)
        if not pr_number:
            log_hook_execution(
                "similar-pattern-search",
                "approve",
                "skipped: could not extract PR number",
            )
            print(json.dumps(result))
            return

        # Get PR diff
        diff = get_pr_diff(pr_number)
        if not diff:
            log_hook_execution(
                "similar-pattern-search",
                "approve",
                f"skipped: could not get diff for PR #{pr_number}",
            )
            print(json.dumps(result))
            return

        # Get changed files to exclude from search
        changed_files = get_changed_files(pr_number)

        # Extract function patterns from diff
        patterns = extract_function_patterns(diff)
        if not patterns:
            log_hook_execution(
                "similar-pattern-search",
                "approve",
                f"skipped: no patterns extracted from PR #{pr_number}",
            )
            print(json.dumps(result))
            return

        # Search for each pattern
        limited_patterns = list(patterns)[:5]  # Limit to 5 patterns
        pattern_results = {}
        for pattern in limited_patterns:
            results = search_pattern_in_codebase(pattern, changed_files)
            if results:
                pattern_results[pattern] = results

        if pattern_results:
            result["systemMessage"] = format_info_message(pattern_results)
            log_hook_execution(
                "similar-pattern-search",
                "approve",
                f"found similar patterns for PR #{pr_number}",
                {
                    "patterns": list(pattern_results.keys()),
                    "total_matches": sum(len(r) for r in pattern_results.values()),
                },
            )
        else:
            log_hook_execution(
                "similar-pattern-search",
                "approve",
                f"no similar patterns found for PR #{pr_number}",
                {"patterns_checked": limited_patterns},
            )

    except Exception as e:
        print(f"[similar-pattern-search] Error: {e}", file=sys.stderr)
        log_hook_execution("similar-pattern-search", "approve", f"error: {e}")

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/similar_pattern_search.ts
#!/usr/bin/env bun
/**
 * PRマージ後にコードベース内の類似パターンを検索し修正漏れを防ぐ。
 *
 * Why:
 *   共通パターン（json.dumps等）を修正する際、同様のパターンが他ファイルに
 *   存在すると修正漏れが発生する。マージ後に自動検索して警告する。
 *
 * What:
 *   - PRマージ成功後（PostToolUse:Bash）に発火
 *   - PR diffから関数呼び出しパターンを抽出
 *   - 変更されたファイル以外で同パターンを検索
 *   - 見つかった場合はsystemMessageで通知
 *
 * State:
 *   - reads: GitHub API (PR diff, changed files)
 *
 * Remarks:
 *   - 非ブロック型（情報提供のみ）
 *   - duplicate-issue-checkはIssue重複、本フックはコードパターン重複
 *   - 一般的すぎる関数（print, len等）はCOMMON_FUNCTIONSで除外
 *
 * Changelog:
 *   - silenvx/dekita#2103: フック追加（Issue #2054/2065の再発防止）
 *   - silenvx/dekita#3162: TypeScriptに移植
 */

import { execFileSync, execSync } from "node:child_process";
import { TIMEOUT_LIGHT, TIMEOUT_MEDIUM } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { getToolResult, parseHookInput } from "../lib/session";

const HOOK_NAME = "similar-pattern-search";

// Maximum number of results to display
const MAX_RESULTS = 5;

// Patterns to exclude from search
const EXCLUDE_PATTERNS = [
  "*.pyc",
  "__pycache__",
  "node_modules",
  ".git",
  "*.min.js",
  "*.min.css",
  "pnpm-lock.yaml",
  "package-lock.json",
];

// Common functions to exclude (too generic)
export const COMMON_FUNCTIONS = new Set([
  "print",
  "len",
  "str",
  "int",
  "float",
  "list",
  "dict",
  "set",
  "tuple",
  "range",
  "enumerate",
  "zip",
  "map",
  "filter",
  "sorted",
  "reversed",
  "open",
  "type",
  "isinstance",
  "hasattr",
  "getattr",
  "setattr",
  "self",
  "super",
  "return",
  "if",
  "for",
  "while",
  "with",
  "assert",
  "raise",
  "except",
  "import",
  "from",
  "class",
  "def",
  "async",
  "await",
  "lambda",
  "get",
  "add",
  "remove",
  "pop",
  "append",
  "extend",
  "update",
  "items",
  "keys",
  "values",
  "join",
  "split",
  "strip",
  "replace",
  "format",
  "lower",
  "upper",
  "startswith",
  "endswith",
  "find",
  "index",
  "count",
]);

export interface SearchResult {
  file: string;
  line: string;
  content: string;
}

/**
 * Check if the command is a PR merge command.
 */
export function isPrMergeCommand(command: string): boolean {
  return command.includes("gh pr merge");
}

/**
 * Check if merge was successful.
 */
export function isMergeSuccess(exitCode: number, output: string): boolean {
  if (exitCode !== 0) {
    return false;
  }

  // Check for merge success patterns
  const successPatterns = [/merged/i, /pull request.*merged/i];
  const failurePatterns = [/failed/i, /error/i, /not mergeable/i];

  // Check for explicit failure
  for (const pattern of failurePatterns) {
    if (pattern.test(output)) {
      return false;
    }
  }

  // Check for explicit success
  for (const pattern of successPatterns) {
    if (pattern.test(output)) {
      return true;
    }
  }

  // Default: assume success if exit code is 0
  return true;
}

/**
 * Extract PR number from merge command or current branch PR.
 */
function extractPrNumber(command: string): number | null {
  // Match patterns like: gh pr merge 123, gh pr merge #123
  const match = command.match(/gh\s+pr\s+merge\s+.*?#?(\d+)/);
  if (match) {
    return Number.parseInt(match[1], 10);
  }

  // If no PR number in command, get PR for current branch
  try {
    const result = execSync("gh pr view --json number", {
      encoding: "utf-8",
      timeout: TIMEOUT_LIGHT * 1000,
    });
    const data = JSON.parse(result);
    return data.number ?? null;
  } catch {
    return null;
  }
}

/**
 * Get the diff of a PR.
 */
function getPrDiff(prNumber: number): string | null {
  try {
    const result = execSync(`gh pr diff ${prNumber}`, {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
    });
    return result;
  } catch {
    return null;
  }
}

/**
 * Get list of files changed in the PR.
 */
function getChangedFiles(prNumber: number): string[] {
  try {
    const result = execSync(`gh pr view ${prNumber} --json files`, {
      encoding: "utf-8",
      timeout: TIMEOUT_LIGHT * 1000,
    });
    const data = JSON.parse(result);
    const files = data.files ?? [];
    return files.map((f: { path?: string }) => f.path ?? "").filter(Boolean);
  } catch {
    return [];
  }
}

/**
 * Extract function call patterns from diff.
 *
 * Focuses on added/modified lines (lines starting with +).
 */
export function extractFunctionPatterns(diff: string): Set<string> {
  const patterns = new Set<string>();

  // Pattern for function calls
  const funcPattern =
    /(?<!def )(?<!class )\b([a-zA-Z_][a-zA-Z0-9_]*(?:\.[a-zA-Z_][a-zA-Z0-9_]*)*)\s*\(/g;

  for (const line of diff.split("\n")) {
    // Focus on added/modified lines
    if (!line.startsWith("+")) {
      continue;
    }
    // Skip diff headers
    if (line.startsWith("+++")) {
      continue;
    }

    // Extract function calls
    for (const match of line.matchAll(funcPattern)) {
      const fullMatch = match[1];
      if (!fullMatch) continue;
      // Get last part for method calls
      const funcName = fullMatch.split(".").pop()?.toLowerCase() ?? "";

      // Skip common functions
      if (!COMMON_FUNCTIONS.has(funcName)) {
        patterns.add(fullMatch);
      }
    }
  }

  return patterns;
}

/**
 * Search for a pattern in the codebase using ripgrep.
 */
function searchPatternInCodebase(pattern: string, excludeFiles: string[]): SearchResult[] {
  const results: SearchResult[] = [];

  // Build exclude arguments
  const excludeArgs: string[] = [];
  for (const excl of EXCLUDE_PATTERNS) {
    excludeArgs.push("-g", `!${excl}`);
  }
  for (const f of excludeFiles) {
    excludeArgs.push("-g", `!${f}`);
  }

  try {
    // Escape pattern for regex
    const escapedPattern = `${pattern.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")}\\s*\\(`;

    const args = [
      "--line-number",
      "--no-heading",
      "--max-count",
      "10",
      ...excludeArgs,
      escapedPattern,
    ];

    // Use execFileSync to avoid shell interpretation of regex special chars
    const result = execFileSync("rg", args, {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
      cwd: process.cwd(),
    });

    if (result.trim()) {
      for (const line of result.trim().split("\n").slice(0, MAX_RESULTS)) {
        // Parse rg output: file:line:content
        const parts = line.split(":");
        if (parts.length >= 3) {
          results.push({
            file: parts[0],
            line: parts[1],
            content: parts.slice(2).join(":").trim().slice(0, 80),
          });
        }
      }
    }
  } catch {
    // rg not available or no matches - return empty
  }

  return results;
}

/**
 * Format the informational message.
 */
export function formatInfoMessage(patternResults: Map<string, SearchResult[]>): string {
  const lines = [
    "🔍 **修正漏れの可能性があります**",
    "",
    "PRで変更されたパターンと類似のコードが他のファイルにあります:",
    "",
  ];

  for (const [pattern, results] of patternResults) {
    lines.push(`**\`${pattern}\`**:`);
    for (const r of results.slice(0, 3)) {
      lines.push(`  - \`${r.file}:${r.line}\` - ${r.content}`);
    }
    if (results.length > 3) {
      lines.push(`  - ... 他 ${results.length - 3} 件`);
    }
    lines.push("");
  }

  lines.push("同様の修正が必要ないか確認してください。");

  return lines.join("\n");
}

async function main(): Promise<void> {
  const result: { continue: boolean; systemMessage?: string } = { continue: true };
  let sessionId: string | undefined;

  try {
    const hookInput = await parseHookInput();
    sessionId = hookInput.session_id;
    if (!hookInput) {
      console.log(JSON.stringify(result));
      return;
    }

    const toolName = hookInput.tool_name ?? "";
    if (toolName !== "Bash") {
      console.log(JSON.stringify(result));
      return;
    }

    const toolInput = (hookInput.tool_input ?? {}) as Record<string, unknown>;
    const command = (toolInput.command as string) ?? "";

    if (!isPrMergeCommand(command)) {
      console.log(JSON.stringify(result));
      return;
    }

    const toolOutput = (hookInput.tool_output as string) ?? "";
    const rawResult = getToolResult(hookInput);
    const toolResult =
      typeof rawResult === "object" && rawResult ? (rawResult as Record<string, unknown>) : {};
    const exitCode = (toolResult.exit_code as number) ?? 0;

    if (!isMergeSuccess(exitCode, toolOutput)) {
      console.log(JSON.stringify(result));
      return;
    }

    const prNumber = extractPrNumber(command);
    if (!prNumber) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        "skipped: could not extract PR number",
        undefined,
        { sessionId },
      );
      console.log(JSON.stringify(result));
      return;
    }

    // Get PR diff
    const diff = getPrDiff(prNumber);
    if (!diff) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `skipped: could not get diff for PR #${prNumber}`,
        undefined,
        { sessionId },
      );
      console.log(JSON.stringify(result));
      return;
    }

    // Get changed files to exclude from search
    const changedFiles = getChangedFiles(prNumber);

    // Extract function patterns from diff
    const patterns = extractFunctionPatterns(diff);
    if (patterns.size === 0) {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `skipped: no patterns extracted from PR #${prNumber}`,
        undefined,
        { sessionId },
      );
      console.log(JSON.stringify(result));
      return;
    }

    // Search for each pattern (limit to 5)
    const limitedPatterns = Array.from(patterns).slice(0, 5);
    const patternResults = new Map<string, SearchResult[]>();

    for (const pattern of limitedPatterns) {
      const results = searchPatternInCodebase(pattern, changedFiles);
      if (results.length > 0) {
        patternResults.set(pattern, results);
      }
    }

    if (patternResults.size > 0) {
      result.systemMessage = formatInfoMessage(patternResults);
      let totalMatches = 0;
      for (const results of patternResults.values()) {
        totalMatches += results.length;
      }
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `found similar patterns for PR #${prNumber}`,
        {
          patterns: Array.from(patternResults.keys()),
          total_matches: totalMatches,
        },
        { sessionId },
      );
    } else {
      await logHookExecution(
        HOOK_NAME,
        "approve",
        `no similar patterns found for PR #${prNumber}`,
        {
          patterns_checked: limitedPatterns,
        },
        { sessionId },
      );
    }
  } catch (error) {
    console.error(`[${HOOK_NAME}] Error: ${formatError(error)}`);
    await logHookExecution(HOOK_NAME, "approve", `error: ${formatError(error)}`, undefined, {
      sessionId,
    });
  }

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
